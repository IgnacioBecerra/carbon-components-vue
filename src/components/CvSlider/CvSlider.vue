<template>
  <div :class="`cv-slider ${carbonPrefix}--form-item`">
    <label
      v-show="label"
      :id="labelId"
      :for="cvId"
      :class="[
        `${carbonPrefix}--label`,
        { [`${carbonPrefix}--label--disabled`]: disabled },
      ]"
    >
      {{ label }}
    </label>

    <div :class="`${carbonPrefix}--slider-container`">
      <span :class="`${carbonPrefix}--slider__range-label`">
        {{ internalMinLabel }}
      </span>

      <div
        :class="[
          `${carbonPrefix}--slider`,
          { [`${carbonPrefix}--slider--disabled`]: disabled },
        ]"
        data-slider
        data-slider-input-box="#slider-input-box"
      >
        <div
          ref="track"
          :class="`${carbonPrefix}--slider__track`"
          @click="onTrackClick"
        ></div>

        <div
          :class="`${carbonPrefix}--slider__filled-track`"
          :style="`width: ${percentage};`"
        ></div>

        <div
          ref="thumb"
          :class="[
            `${carbonPrefix}--slider__thumb`,
            {
              [`${carbonPrefix}--slider__thumb--clicked`]: animateClick,
            },
          ]"
          tabindex="0"
          :aria-labelledby="labelId"
          :style="`left: ${percentage};`"
          @keydown.up.right.prevent="onUp"
          @keydown.down.left.prevent="onDown"
          @mousedown="onStartDrag"
        ></div>

        <input
          :id="cvId"
          ref="range"
          :class="`${carbonPrefix}--slider__input`"
          type="range"
          :step="step"
          :min="min"
          :max="max"
        />
      </div>

      <span :class="`${carbonPrefix}--slider__range-label`">
        {{ internalMaxLabel }}
      </span>

      <input
        ref="inputBox"
        v-model="internalValue"
        type="string"
        :class="[
          `${carbonPrefix}--text-input ${carbonPrefix}--slider-text-input`,
          { [`${carbonPrefix}--text-input--light`]: isLight },
        ]"
        :placeholder="min"
        :disabled="disabled || null"
        @change="onChange"
        @keydown.up.prevent="onUp"
        @keydown.down.prevent="onDown"
      />
    </div>
  </div>
</template>

<script setup>
import { computed, ref, onMounted, nextTick, watch } from 'vue';
import { carbonPrefix } from '../../global/settings';
import { props as propsCvId, useCvId } from '../../use/cvId';
import { useIsLight, props as propsTheme } from '../../use/cvTheme';

const props = defineProps({
  modelValue: {
    type: String,
    default: undefined,
  },
  value: {
    type: String,
    default: undefined,
  },
  disabled: Boolean,
  label: {
    type: String,
    default: undefined,
  },
  // NOTE: It is not safe to rely on Numbers for non-integer steps
  min: {
    type: String,
    default: '0',
  },
  max: {
    type: String,
    default: '100',
  },
  minLabel: {
    type: String,
    default: undefined,
  },
  maxLabel: {
    type: String,
    default: undefined,
  },
  step: {
    type: String,
    default: '1',
  },
  stepMultiplier: {
    type: String,
    default: '4',
    validator(val) {
      if (val.length) {
        let intMultiplier = parseInt(val);
        if (isNaN(intMultiplier) || intMultiplier < 1) {
          console.warn('cv-slider: multiplier must be >= 1');
          return false;
        }
      }
      return true;
    },
  },
  ...propsTheme,
  ...propsCvId,
});

const range = ref(null);
const thumb = ref(null);
const track = ref(null);
const DefaultRangeMin = 0;
const DefaultRangeMax = 100;
const DefaultValue = 0;
const StepNaNReplacement = 0;
const StepDefaultValue = 1;
const DefaultMultiplier = 4;

const emit = defineEmits(['update:modelValue', 'change', 'update:value']);

const cvId = useCvId(props, true);
const isLight = useIsLight(props);
const internalValue = ref('');
const animateClick = ref(false);
const isDragging = ref(false);
const dragStartX = ref(0);
const dragStartValue = ref(0);
const percentage = ref('0%');

const labelId = computed(() => `${cvId.value}-label`);

const internalMinLabel = computed(() => props.minLabel ?? getMin());

const internalMaxLabel = computed(() => props.maxLabel ?? getMax());

const rangePropsObserver = computed(() => ({
  min: props.min,
  max: props.max,
  step: props.step,
}));

const internalMultiplier = computed(() => {
  let intMultiplier = parseInt(props.stepMultiplier);
  // default to 4 fro multiplier
  return isNaN(intMultiplier) ? DefaultMultiplier : Math.max(intMultiplier, 1);
});

const getRangeAttributeValue = (
  attributeName,
  defaultForNaN,
  defaultReturnValue
) => {
  if (range?.value) {
    const val = parseFloat(range.value[attributeName]);
    return isNaN(val) ? defaultForNaN : val;
  }
  return defaultReturnValue ?? defaultForNaN;
};

const getMin = () => {
  return getRangeAttributeValue('min', DefaultRangeMin);
};

const getMax = () => {
  return getRangeAttributeValue('max', DefaultRangeMax);
};

const getValue = () => {
  const middleValue = (getMax() + getMin()) / 2;
  return getRangeAttributeValue('value', DefaultValue, middleValue);
};

const getStep = () => {
  return getRangeAttributeValue('step', StepNaNReplacement, StepDefaultValue);
};

const setValue = newValue => {
  if (props.disabled) return;

  range.value.value = newValue;
  internalValue.value = range.value.value;

  percentage.value = `${
    ((internalValue.value - getMin()) * 100) / (getMax() - getMin())
  }%`;

  const { value: rangeValue } = range.value;
  emit('update:modelValue', rangeValue);
  emit('update:value', rangeValue);
  emit('change', rangeValue);
};

const onChange = () => {
  let newValue = internalValue.value.length
    ? parseFloat(internalValue.value)
    : getMin();
  setValue(newValue);
};

const onStartDrag = ev => {
  document.body.addEventListener('mousemove', onDrag);
  document.body.addEventListener('mouseup', onStopDrag);

  dragStartX.value = ev.clientX;
  dragStartValue.value = getValue();
  isDragging.value = true;
};

const onDrag = ev => {
  if (isDragging.value) {
    // percentage change
    let newValue = (ev.clientX - dragStartX.value) / track.value.offsetWidth;
    // uncapped new value
    newValue = dragStartValue.value + (getMax() - getMin()) * newValue;
    setValue(newValue);
  }
};

const onStopDrag = () => {
  isDragging.value = false;
  document.body.removeEventListener('mousemove', onDrag);
  document.body.removeEventListener('mouseup', onStopDrag);
};

const onTrackClick = ev => {
  const afterAnimate = ev => {
    if (ev.propertyName === 'left') {
      animateClick.value = false;
      thumb.value.removeEventListener('transitionend', afterAnimate);
    }
  };

  let newValue = ev.offsetX / track.value.offsetWidth;
  const min = getMin();
  newValue = (getMax() - min) * newValue + min;
  thumb.value.addEventListener('transitionend', afterAnimate);
  animateClick.value = true;
  setValue(newValue);
};

const onUpDown = (ev, isUp = true) => {
  let curValue =
    ev.target.type === 'number' ? parseFloat(ev.target.value) : getValue();
  const step = getStep();
  const progressValue = ev.shiftKey ? internalMultiplier.value * step : step;
  const direction = isUp ? 1 : -1;
  let newValue = curValue + progressValue * direction;
  setValue(newValue);
};

const onUp = ev => {
  onUpDown(ev);
};

const onDown = ev => {
  onUpDown(ev, false);
};

onMounted(() => {
  range.value.value = props.value ?? props.modelValue;
  internalValue.value = range.value.value;
  percentage.value = `${
    ((internalValue.value - getMin()) * 100) / (getMax() - getMin())
  }%`;
});

watch(
  () => props.modelValue,
  modelValue => {
    setValue(modelValue);
  }
);

watch(
  () => props.value,
  value => {
    setValue(value);
  }
);

watch(rangePropsObserver, () => {
  nextTick(() => {
    setValue(internalValue.value);
  });
});
</script>
