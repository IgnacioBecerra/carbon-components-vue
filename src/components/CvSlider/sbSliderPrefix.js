/* istanbul ignore next */
export const sbSliderPrefix = 'CvSlider';
