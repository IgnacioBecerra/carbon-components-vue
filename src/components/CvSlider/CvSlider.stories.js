import {
  sbCompPrefix,
  storyParametersObject,
} from '../../global/storybook-utils';
import { sbSliderPrefix } from './sbSliderPrefix';
import { CvSlider } from '.';

export default {
  title: `${sbCompPrefix}/${sbSliderPrefix}/CvSlider`,
  component: CvSlider,
  argTypes: {
    // TODO
  },
};

const template = `<cv-slider v-bind="args" />`;
const Template = args => {
  return {
    components: { CvSlider },
    setup: () => ({ args }),
    template,
  };
};

export const Default = Template.bind({});
Default.parameters = storyParametersObject(
  Default.parameters,
  template,
  Default.args
);
