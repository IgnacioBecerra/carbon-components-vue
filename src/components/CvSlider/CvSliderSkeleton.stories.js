import {
  sbCompPrefix,
  storyParametersObject,
} from '../../global/storybook-utils';
import { sbSliderPrefix } from './sbSliderPrefix';
import { CvSliderSkeleton } from '.';

export default {
  title: `${sbCompPrefix}/${sbSliderPrefix}/CvSliderSkeleton`,
  component: CvSliderSkeleton,
};

const template = `<cv-slider-skeleton />`;
const Template = () => {
  return {
    components: { CvSliderSkeleton },
    template,
  };
};

export const Default = Template.bind({});
Default.parameters = storyParametersObject(
  Default.parameters,
  template,
  Default.args
);
