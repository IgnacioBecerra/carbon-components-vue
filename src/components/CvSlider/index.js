import { default as CvSlider } from './CvSlider.vue';
import { default as CvSliderSkeleton } from './CvSliderSkeleton.vue';

export { CvSlider, CvSliderSkeleton };
export default CvSlider;
