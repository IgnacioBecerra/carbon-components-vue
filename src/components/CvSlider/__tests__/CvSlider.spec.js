import { shallowMount, mount } from '@vue/test-utils';
import { nextTick } from 'vue';
import { carbonPrefix } from '../../../global/settings';
import CvSlider from '../CvSlider.vue';

const DefaultValue = '7';
const DefaultNewValue = '34';
const DefaultPropsValue = {
  props: {
    value: DefaultValue,
  },
};
const DefaultPropsModelValue = {
  props: {
    modelValue: DefaultValue,
  },
};

const HTMLNodesClasses = {
  track: `.${carbonPrefix}--slider__track`,
  valueInput: `.${carbonPrefix}--text-input`,
  thumb: `.${carbonPrefix}--slider__thumb`,
};

describe('CvSlider', () => {
  it('should match snapshot', async () => {
    const wrapper = shallowMount(CvSlider, {
      props: {
        id: 'abc',
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  test.each([
    ['keydown.up', '8'],
    ['keydown.down', '6'],
  ])('handles %s press correctly', async (keyCode, expectedValue) => {
    const wrapper = shallowMount(CvSlider, DefaultPropsValue);
    const input = wrapper.find(HTMLNodesClasses.valueInput);
    await input.trigger(keyCode);
    await nextTick();

    expect(wrapper.emitted('change')[0][0]).toEqual(expectedValue);
  });

  it('handles drag correctly', async () => {
    const InitialValue = 1;
    const wrapper = shallowMount(CvSlider, {
      props: {
        value: InitialValue.toString(),
      },
    });
    const thumb = wrapper.find(HTMLNodesClasses.thumb);
    await thumb.trigger('mousedown');
    document.body.dispatchEvent(new MouseEvent('mousemove', { clientX: 100 }));
    document.body.dispatchEvent(new MouseEvent('mouseup'));

    await nextTick();
    const emittedValue = wrapper.emitted('change')[0][0];

    // Assert that value has been updated due to drag
    expect(Number.isNaN(emittedValue)).toBe(false);
    expect(Number(emittedValue)).toBeGreaterThan(InitialValue);
  });

  it('emits v-model event when modelValue prop changes', async () => {
    const wrapper = shallowMount(CvSlider, DefaultPropsModelValue);

    await wrapper.setProps({ modelValue: DefaultNewValue });

    expect(wrapper.emitted('update:modelValue')[0][0]).toEqual(DefaultNewValue);
  });

  it('emits change event when value prop changes', async () => {
    const wrapper = shallowMount(CvSlider, DefaultPropsValue);

    await wrapper.setProps({ value: DefaultNewValue });

    expect(wrapper.emitted('change')[0][0]).toEqual(DefaultNewValue);
  });

  test.each([
    ['step', '12', DefaultValue],
    ['min', '10', '10'],
    ['max', '6', '6'],
  ])(
    'emits change event when %s prop changes',
    async (propName, propValue, expectedValue) => {
      const wrapper = shallowMount(CvSlider, DefaultPropsValue);

      await wrapper.setProps({ [propName]: propValue });
      await nextTick();

      expect(wrapper.emitted('change')[0][0]).toEqual(expectedValue);
    }
  );

  it('handles value change from input box correctly', async () => {
    const wrapper = shallowMount(CvSlider);
    const inputBox = wrapper.find(HTMLNodesClasses.valueInput);
    await inputBox.setValue(DefaultNewValue);
    await inputBox.trigger('change');

    expect(wrapper.emitted('change')[0][0]).toEqual(DefaultNewValue);
  });

  it('respects stepMultiplier when adjusting value', async () => {
    const wrapper = shallowMount(CvSlider, {
      props: {
        step: '3',
        stepMultiplier: '5',
        value: '1',
      },
    });

    let inputBox = wrapper.find(HTMLNodesClasses.valueInput);

    await inputBox.trigger('keydown.up', { shiftKey: true });
    await nextTick();
    await inputBox.trigger('keydown.down', { shiftKey: true });
    await nextTick();

    expect(wrapper.emitted('change')[0][0]).toEqual('16');
    expect(wrapper.emitted('change')[1][0]).toEqual('1');
  });

  // it.only('updates value when track is clicked', async () => {
  //   const wrapper = shallowMount(CvSlider);

  //   const track = wrapper.find(HTMLNodesClasses.track);
  //   await track.trigger('click', { offsetX: 50 });

  //   // internalValue should be updated based on click position
  //   expect(wrapper.vm.internalValue).toBeGreaterThan(0);
  // });

  // it('handles arrow key press on thumb correctly', async () => {
  //   const wrapper = mount(CvSlider);
  //   const thumb = wrapper.find('.cv-slider__thumb');

  //   await thumb.trigger('keydown.up');
  //   expect(wrapper.vm.internalValue).toBeGreaterThan(0);

  //   await thumb.trigger('keydown.down');
  //   expect(wrapper.vm.internalValue).toBe(0);
  // });
});
