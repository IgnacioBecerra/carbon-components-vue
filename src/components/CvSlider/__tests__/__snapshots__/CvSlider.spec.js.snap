// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CvSlider should match snapshot 1`] = `
<div class="cv-slider bx--form-item"><label id="abc-label" for="abc" class="bx--label" style="display: none;"></label>
  <div class="bx--slider-container"><span class="bx--slider__range-label">0</span>
    <div class="bx--slider" data-slider="" data-slider-input-box="#slider-input-box">
      <div class="bx--slider__track"></div>
      <div class="bx--slider__filled-track" style="width: 0%;"></div>
      <div class="bx--slider__thumb" tabindex="0" aria-labelledby="abc-label" style="left: 0%;"></div><input id="abc" class="bx--slider__input" type="range" step="1" min="0" max="100">
    </div><span class="bx--slider__range-label">100</span><input type="string" class="bx--text-input bx--slider-text-input" placeholder="0">
  </div>
</div>
`;
